packages/nextjs/utils/scaffold-stark/contract.ts

```diff
@@ -411,238 +420,444 @@ function tryParsingParamReturnObject(fn: (x: any) => {}, param: any) {
-    } else if (isCairoArray(paramType)) {
-      return tryParsingParamReturnObject((param) => {
-        const genericType = parseGenericType(paramType)[0];
-        return genericType
-          ? //@ts-ignore
-            param.map((item) => parseParamWithType(genericType, item, isRead))
-          : param;
-      }, param);
-    } else if (isCairoOption(paramType)) {
-      //@ts-ignore
-      return tryParsingParamReturnObject((x) => {
-        const option = x as CairoOption<any>;
-        return option.isNone()
-          ? "None"
-          : `Some(${parseParamWithType(paramType.split("<").pop()!, option.unwrap(), isRead)})`;
-      }, param);
-    } else if (isCairoResult(paramType)) {
-      //@ts-ignore
-      return tryParsingParamReturnObject((x) => {
-        const result = x as CairoResult<any, any>;
-        const [ok, error] = parseGenericType(paramType);
-        return result.isOk()
-          ? `Ok(${parseParamWithType(ok, result.unwrap(), isRead)})`
-          : `Err(${parseParamWithType(error, result.unwrap(), isRead)})`;
-      }, param);
-    } else if (isCairoU256(paramType)) {
-      return tryParsingParamReturnObject(uint256.uint256ToBN, param);
-    } else if (isCairoByteArray(paramType)) {
-      return tryParsingParamReturnObject(byteArray.stringFromByteArray, param);
-    } else if (isCairoContractAddress(paramType)) {
-      return tryParsingParamReturnObject(validateAndParseAddress, param);
-    } else if (isCairoFelt(paramType)) {
-      return feltToHex(param);
-    } else if (isCairoBool(paramType)) {
-      return typeof param === "boolean"
-        ? param
-        : (param as string).startsWith("0x0")
-          ? "false"
-          : "true";
-    } else if (isCairoBytes31(paramType)) {
-      return tryParsingParamReturnObject(
-        (x: bigint) => `0x${x.toString(16)}`,
-        param,
-      );
-    } else if (isCairoInt(paramType)) {
-      return tryParsingParamReturnObject(
-        (x) => (typeof x === "bigint" ? Number(x) : parseInt(x, 16)),
-        param,
-      );
-    } else if (isCairoBigInt(paramType)) {
-      return tryParsingParamReturnObject((x) => BigInt(x), param);
-    } else {
-      return tryParsingParamReturnObject((x) => x, param);
-    }
-  } else {
-    if (isCairoTuple(paramType)) {
-      return tryParsingParamReturnValues(
-        (x) => stringToObjectTuple(x, paramType),
-        param,
-      );
-    } else if (isCairoArray(paramType)) {
+const decodeParamsWithType = (paramType: string, param: any): unknown => {
+  const isRead = true;
+  if (isCairoTuple(paramType)) {
+    return objectToCairoTuple(param, paramType);
+  } else if (isCairoArray(paramType)) {
+    return tryParsingParamReturnObject((param) => {
       const genericType = parseGenericType(paramType)[0];
+      return genericType
+        ? //@ts-ignore
+          param.map((item) => parseParamWithType(genericType, item, isRead))
+        : param;
+    }, param);
+  } else if (isCairoOption(paramType)) {
```

Comment:
wouldnt this change from `tryParsingParamReturnObject` to `tryParsingParamReturnValues` work for all the other places where `tryParsingParamReturnObject` is located, by just setting the v3 flag to true. @metalboyrick

Reply:
this flag is for encoding data for read purposes. new `starknet-react` hook `useReadContract` doesn't have an optional parsing flag like in v2, so all args have to be parsed as-is or be converted into `starknet-js` format. But writing would still need to parse into rawArgs.